### ISSUE TITLE
How to properly resume decoder training?

### ISSUE THREAD

**Issue by heumchri** [Eromera/erfnet_pytorch#28]
Hi, 

I am trying to retrain the model on my own on the cityscapes dataset, but only using 2 classes. The encoder training works fine, but I have problems with decoder training which I think are caused by not attaching the trained encoder to the model properly.

As far as i understand, there are two possibilities for decoder training. 

**1. Use the encoder pretrained on imagenet**

For this I used the following commands from the documentation

`python main_binary.py --savedir erfnet_training1 --datadir /home/datasets/cityscapes/ --num-epochs 150 --batch-size 6 --decoder --pretrainedEncoder "../trained_models/erfnet_encoder_pretrained.pth.tar"`
and
`python main_binary.py --savedir erfnet_training1 --datadir /home/datasets/cityscapes/ --num-epochs 150 --batch-size 6 --decoder --pretrainedEncoder "../trained_models/erfnet_encoder_pretrained.pth.tar" --resume`

I have not trained the model for all epochs, but the intermediate result looks fine (best Val-IoU after 85 epochs: 0.9495)

**2. Use an encoder trained on cityscapes**

The encoder training worked fine and resulted in a Val-IoU of 0.9471.
However, I couldn't find any documentation on how to attach the pretrained encoder for decoder training.
From the code, I got that --pretrainedEncoder flag is only for the imagenet encoder, and that I should use --state, so I used

`python main_binary.py --savedir erfnet_training2 --datadir /home/datasets/cityscapes/ --num-epochs 150 --batch-size 6 --decoder --state "../save/erfnet_training2/model_best_enc.pth.tar"`
and since the code says to only use --state for initializing:
`python main_binary.py --savedir erfnet_training2 --datadir /home/datasets/cityscapes/ --num-epochs 150 --batch-size 6 --decoder  --resume`

However, this didn't work out and finished with a best Val-IoU of only 0.9441, so the whole network performs worse than just the encoder alone.
For comparison, I also tested training the decoder without initializing an encoder, which resulted in Val-IoU= 0.9461.

So what is the correct way of training the decoder after finished encoder training?
Especially the arguments in combination with the --resume flag, since I cannot train the model in one go due to hardware availability.

Thank You for your answer.

**Comment by Eromera** [comment]
Hi!

Sorry for the late reply. The way that you mention in the section 2. should be alright. --decoder + --state for init and --decoder + --resume for resuming a snapshot. The fact that you are getting a worse result in the decoder training than encoder is reasonable since encoder evaluates with images downsampled by 1/8. The fact that training the full network without pretraining the encoder results in a better IoU is a bit more strange but this might also happen if you train a sufficiently large number of epochs (like 150 that you are using) since this gives enough time to the first layers weights to receive enough updates.

**Comment by InstantWindy** [comment]
@heumchri,Did you test the validation set mean IOU at 94.71% on the cityscape dataset?I used this model to test the mean iou of the data set of cityscapes is not so high, only more than 70%.Thank you for your reply!

**Comment by Eromera** [comment]
@InstantWindy I think he is getting 94.71% using only **2 classes**.

**Comment by InstantWindy** [comment]
How do you read data sets? Your dataset file simply describes the reading of the cityscape dataset, but does not involve the encoding of the tag. The cityscape dataset is a total of 19 classes. Your file contains 20 classes. I'm currently trying to implement it with tensorflow。Thank you!

**Comment by Eromera** [comment]
What do you mean it does not involve the encoding of the tag?
Cityscapes is 19 classes + the ignore label, thats why there are 20. 
If you are interested in a tensorflow implementation maybe you can try [this one](https://github.com/ronrest/erfnet_segmentation) (I haven't tried it yet)

**Comment by InstantWindy** [comment]
Yeah, thank you very much! The cityscapes dataset labels have a pixel value range of 0-33, how do you convert the labels' pixel values to 0-20?

**Comment by heumchri** [comment]
Hi @Eromera ,

thank you for your answer. 

I did some further training in the mean time with the following results.
_Val-IoU_ is the accuracy in the best(_encoder).txt, so the mean of the 2 classes in the size the model is working in (1024x512 or 128x64)
_road IoU_ is measured using the official cityscapes scripts on the val set on outputs upsampled by bilinear interpolation to the original cityscapes size (2048x1024)

Best encoder:
Val-IoU:    0.9473
road IoU:  **0.963**

Best decoder with encoder pretrained on imagenet:
Val-IoU:    0.9501
road IoU:  **0.970**

Best decoder with encoder pretrained on cityscapes (the one mentioned above):
Val-IoU:    0.9455
road IoU:  **0.962**

Best decoder without pretrained encoder:
Val-IoU:    0.9503
road IoU:  **0.967**

For comparison:
Best decoder on 19 classes provided by you:
Val-IoU:    0.9503
road IoU:  **0.976**
mIOU over all 19 classes in 2048x1024: 0.772

There are still some questions i have about these results. 

1. The best decoder is still worse than just the encoder alone, even when evaluating on the same image size (**0.962** vs **0.963**) 
I understand that the decoder won't achieve a huge performance gain on the road class because roads are large and with few sharp edges, so the low resolution of the encoder already gives a good approximation. The decoder is much more important for small classes like traffic signs or lights.
However, on the road class, the decoder should at least achieve a little accuracy gain, right?

2. The initial goal of just using 2 classes for training was to achieve a better result on the road segmentation compared to just using your network for 20 classes and ignoring all except the road. 
However, i couldn't even achieve the same accuracy (**0.970** vs **0.976**) and i expected it to be better. Do you have any idea, why that might be the case?
My assumption is that it has something to do with the class weights. 
For 20 classes, you used the following weights:

> if (enc):
        weight[0] = 2.3653597831726	
        weight[1] = 4.4237880706787	
        weight[2] = 2.9691488742828	
        weight[3] = 5.3442072868347	
        weight[4] = 5.2983593940735	
        weight[5] = 5.2275490760803	
        weight[6] = 5.4394111633301	
        weight[7] = 5.3659925460815	
        weight[8] = 3.4170460700989	
        weight[9] = 5.2414722442627	
        weight[10] = 4.7376127243042	
        weight[11] = 5.2286224365234	
        weight[12] = 5.455126285553	
        weight[13] = 4.3019247055054	
        weight[14] = 5.4264230728149	
        weight[15] = 5.4331531524658	
        weight[16] = 5.433765411377	
        weight[17] = 5.4631009101868	
        weight[18] = 5.3947434425354
    else:
        weight[0] = 2.8149201869965	
        weight[1] = 6.9850029945374	
        weight[2] = 3.7890393733978	
        weight[3] = 9.9428062438965	
        weight[4] = 9.7702074050903	
        weight[5] = 9.5110931396484	
        weight[6] = 10.311357498169	
        weight[7] = 10.026463508606	
        weight[8] = 4.6323022842407	
        weight[9] = 9.5608062744141	
        weight[10] = 7.8698215484619	
        weight[11] = 9.5168733596802	
        weight[12] = 10.373730659485	
        weight[13] = 6.6616044044495	
        weight[14] = 10.260489463806	
        weight[15] = 10.287888526917	
        weight[16] = 10.289801597595	
        weight[17] = 10.405355453491	
        weight[18] = 10.138095855713	

   > weight[19] = 0

how did you calculate them? 
Also, the background class has weight 0. Since i only have two classes, i cannot ignore the background because 2 classes are needed for the cross entropy loss. 
For the models mentioned above, I trained them with both classes weighted the same. I also tried with weights = 1/class probability, but the result was not better. 
How would you suggest I calculate the weights when i have only two classes (road, background)?

Also, if you have any other ideas how I can achieve the same or better result than your pretrained model (**0.976**), I would really apreciate it.

Thank you very much for your answer.




@InstantWindy cityscapes has 30 classes, but only 19 are used for the benchmark. Thats why the model is only trained on these classes. You can convert the labels using the official cityscapes scripts:
https://github.com/mcordts/cityscapesScripts
escpecially look at `preparation/createTrainIdLabelImgs.py`

**Comment by InstantWindy** [comment]
@heumchri  But after the conversion, the pixel value is 255. How do you deal with it?Thank you !

**Comment by heumchri** [comment]
@InstantWindy you can specify which values which classes should be assigned in`helpers/labels.py`

**Comment by Eromera** [comment]
@heumchri 1) theoretically it should be better with decoder but not necessarily since you are getting pretty high values (0.96) which are hard to improve, and road is one of those classes that do not need that much resolution as you say. 2) My assumption is that the CNN benefits from multiple classes to improve the performance in other classes. For example, if you set all other classes to background, then during training it cannot benefit from context information (such as knowing that road is below cars or by sidewalks). My only suggestion about the weights would be to use pixel frequency. Maybe you can try to incorporate a few other classes that could be important to detect road.

@InstantWindy as heumchri said, you can specify it in helpers/labels.py, or in the code there is a MyCoTransform that relabels 255 to 19 for the ignore label.

**Comment by heumchri** [comment]
@Eromera Thank you for your input. The lack of context seems reasonable.
For the weights I've tried the inverse class probability. So if class 0 has 2 pixels and class 1 has 3, then weight[0] = 5/2 and weight[1] = 5/3. But this approach didn't result in a better accuracy either.
I assume this is what you mean with pixel frequency?

How did you calculate the weights you use in the code for 20 classes? It doesn't seem to be pixel frequency.

**Comment by Eromera** [comment]
@heumchri Maybe instead of direct inverse you can use inverse of logarithm or similar. The weights in the code are a hardcopy of the ones I got in Torch, you can see the implementation [here](https://github.com/Eromera/erfnet/blob/master/train/model/loss_general.lua), and histclasses is calculated [here](https://github.com/Eromera/erfnet/blob/master/train/data/loadCityscape8bit.lua).

**Comment by InstantWindy** [comment]
@Eromera But I have a question, the cityscapes dataset is a total of 19 classes, then you added a neglected class, but did not train the 19th class during training, so the final predicted output label range is 0-18, but how do you visualize the predicted output  19th class?

**Comment by Eromera** [comment]
Hi @InstantWindy, you can do both. You can train with a neglected class but in that case you will not be able to visualize the neglected (black) class at inference. Or you can train including that 19th class as an extra class which will show during inference.

**Comment by InstantWindy** [comment]
Hi @Eromera,Thank you for your reply. I used the "eval/eval_cityscapes_server.py" script with "--subset test" option . I uploaded the test results as he asked, but there was always Content of zip archive does not match specifications. Not all predictions for all images were found.  Thank you very much！

**Comment by Eromera** [comment]
Hi @InstantWindy, that's weird... How many images do you have in the folder that you are compressing?

**Comment by InstantWindy** [comment]
All the test images. A total of 1524 pictures, the size is 1024 x2048.

**Comment by Eromera** [comment]
Humm test images should be 1525 so it seems you are missing one. Can you check if you have all of them in the original cityscapes folder?

**Comment by InstantWindy** [comment]
Thank you very much, after submitting the results, how long does it take to get results?

**Comment by InstantWindy** [comment]
Did you upload the test image result that contains the 19th class?Or is the test result only 0-18 class?

**Comment by Eromera** [comment]
Hi, it should take a few minutes. You can upload both, since the Cityscapes code will ignore the 19th class. However I think that you will get slightly better results if you submit only 0-18 classes.

**Comment by InstantWindy** [comment]
ok,Thank you very much!What is the batch size you use when testing the speed of the model?When you use the pytorch framework, does the speed of the model relate to the size of batch size?

**Comment by InstantWindy** [comment]
Because I tested at TITAN X  GPU  slower than the paper's.

**Comment by Eromera** [comment]
Hi, the batch size I normally use for test is 1. There is not improvement if you use a larger batch size at test since the GPU is already performing at full capacity, unless you plan on doing multi-gpu.

**Comment by InstantWindy** [comment]
time_val=[]
num=0
mean_time=0
total_time=0
total_imgs=0
fps=[]
for step, (images, labels) in enumerate(loader_val):
    inputs = Variable(images, volatile=True).cuda()
    batch_imgs=inputs.size(0)
    start_time = time.time()
    outputs = net(inputs)
    if (not False):
        torch.cuda.synchronize()    #wait for cuda to finish (cuda is asynchronous!)
    if step!=0:    #first run always takes some time for setup
        fwt = time.time() - start_time
        total_time +=fwt
        time_val.append(fwt)
        mean=sum(time_val) / len(time_val) / batch_imgs
        mean_time +=mean
        num +=1
        total_imgs += batch_imgs
        print ("Forward time per img (b=%d): %.3f (Mean: %.3f)" % (batch_imgs, fwt/batch_imgs, sum(time_val) / len(time_val) / batch_imgs))
        print('Generated segmentations in %s seconds -- %s FPS' % (fwt, batch_imgs/fwt))
        fps.append(batch_imgs/fwt)
    time.sleep(1)   #to avoid overheating the GPU too much

print()
print('average fps:',sum(fps)/len(fps))
print('average fwt:',mean_time/num)

print('Generated segmentations in %s seconds -- %s FPS' % (total_time, total_imgs/total_time)) 

I use the above code to test the speed of the model.I used TITAN X GPU to test the speed of ERFNet and found that it was slower than the paper, because it is possible that the CPU is different, so the test results are biased? Thank you!

**Comment by hydxqing** [comment]
@heumchri Ｈow did you get the value of weight[0] and weight[1]?
In your 2 classes of dataset, does it include includes a class of objects, a class of background?

**Comment by Eromera** [comment]
Hi @InstantWindy the results on the paper are by running the model without batchnorm layers since they can be absorved by the conv layers, and also making sure that the gpu is operating at full power and without any other elements.

@hydxqing You can get these values by calculating the frequencies of these classes in terms of pixels in the dataset
